File: aoc-2021/packages/aoc-go/solutions/day_01.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func partOne() int {
	file, _ := os.Open("../../input/day_01.in")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	count := 0
	scanner.Scan()
	prev, _ := strconv.Atoi(scanner.Text())
	for scanner.Scan() {
		d, _ := strconv.Atoi(scanner.Text())
		if d > prev {
			count++
		}
		prev = d
	}
	return count
}

func partTwo() int {
	content, _ := ioutil.ReadFile("../../input/day_01.in")
	lines := strings.Split(string(content), "\n")
	data := make([]int, len(lines))
	for i, l := range lines {
		data[i], _ = strconv.Atoi(l)
	}
	count := 0
	const window = 3
	currSum := 0
	prevSum := 0
	for i := 0; i < window; i++ {
		prevSum += data[i]
	}

	for i := 1; i < len(data)-2; i++ {
		for j := 0; j < window; j++ {
			currSum += data[i+j]
		}
		if currSum > prevSum {
			count++
		}
		prevSum = currSum
		currSum = 0
	}

	return count
}

func main() {
	fmt.Printf("Answer part one: %d\n", partOne())
	fmt.Printf("Answer part two: %d\n", partTwo())
}

File: aoc-2021/packages/aoc-rust/src/day_01.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

fn part_one() -> i32 {
    let mut data = BufReader::new(File::open("../../input/day_01.in").unwrap())
        .lines()
        .filter_map(|l| l.ok())
        .filter_map(|l| l.parse::<i32>().ok());

    let mut count = 0;
    let mut prev = data.next().unwrap();
    for d in data {
        if d > prev {
            count += 1;
        }
        prev = d;
    }
    count
}

fn part_two() -> i32 {
    let data: Vec<i32> = BufReader::new(File::open("../../input/day_01.in").unwrap())
        .lines()
        .filter_map(|l| l.ok())
        .filter_map(|l| l.parse::<i32>().ok())
        .collect();
    let mut count = 0;
    let mut sum = 0;
    let mut prev_sum = data[0] + data[1] + data[2];
    let win = 3;
    for i in 1..data.len() - 2 {
        for j in 0..win {
            sum += data[i + j as usize];
        }
        if sum > prev_sum {
            count += 1;
        }
        prev_sum = sum;
        sum = 0;
    }
    count
}

pub fn show() {
    println!("- Part one answer: {}", part_one());
    println!("- Part two answer: {}", part_two());
}
